Fix string numbers in to_rgba() and is_color_like()

timhoffm opened the pull request:
## PR Summary

`matplotlib.colors.to_rgba()` and `matplotlib.colors.is_colorlike()` accepted strings representing numbers outside the 0-1 range (converting `'5'` to `(5.0, 5.0, 5.0, 1.0)`). This is a bug and should error out.

Closes #13912.

anntzer commented on lib/matplotlib/colors.py:
```diff
@@ -225,9 +225,14 @@ def _to_rgba_no_colorcycle(c, alpha=None):
+            if not (0 <= c <= 1):
+                raise ValueError(
+                    "String grayscale values must be within 0-1 range")
+            return (float(c),) * 3 + (alpha if alpha is not None else 1.,)
```
no need to convert to float again, `(c, c, c, alpha if ... else 1)` is good enough :)

ImportanceOfBeingErnest commented on lib/matplotlib/colors.py:
```diff
@@ -225,9 +225,14 @@ def _to_rgba_no_colorcycle(c, alpha=None):
+        else:
+            if not (0 <= c <= 1):
+                raise ValueError(
+                    "String grayscale values must be within 0-1 range")
```
All other ValueErrors report back the `orig_c`. Would it make sense to do so here as well?
Possibly:

    "Invalid string grayscale value {!r}. Values must be within 0-1 range.".format(orig_c)

dopplershift commented:
I'm not sure I agree that this is an API change. I'd say it was a bug that this behavior ever worked. IMO, an API change is a change from one intentional behavior to a different intentional behavior. Behavior that was not intended is not part of the API. That's not to say the fix won't be disruptive, but that's the nature of development.

ImportanceOfBeingErnest commented:
It's an API change in the very sense of the definition. You ask the interface a question, like `is_colorlike("5")` and it gives an answer, `True`. Now this is changed to give the answer `False`. Clearly this change is useful and desired, but it's always good to document those changes in the list that exists for that very purpose.

timhoffm commented:
Added an API change note. Even though I'm on the edge of argumenting that this adds more noise than helps. Would we have added an API change note if `is_colorlike('5')` had raised an error instead? Also I doubt, that this could have been used in working code, because using the returned (5., 5., 5., 1.) tuple anywhere in matplotlib would have raised an error there. In that sense, I don't think this a *breaking* API change.

dopplershift commented:
@ImportanceOfBeingErnest By that definition, ***every*** bug fix is an API change. And I don't think that's hyperbole based on your argument.

I'm not trying to die on this hill and if people want to put in an API change note for this, I don't actually care. I'm more confused and surprised that the consensus is in favor of this being an API change.

dstansby commented:
There might be people out there who rely on the behaviour `is_colorlike("5")`, and I think it's better to at least leave a note somewhere to tell them why it's not working any more.
